### PR TITLE
Do not create StatefulSet pods when PVC is being deleted

### PR DESCRIPTION
**Pull request by ymmt2005**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Pod with PVC will not be scheduled if the PVC is being deleted.
This can happen when the PVC has finalizers of storage plugins.

Such a pod becomes pending.  Unfortunately, after the finalizer
finishes and PVC is deleted, the pod remains pending forever.
The StatefulSet controller does nothing for this pending pod.

This commit prevents the StatefulSet controller from creating
such pods when PVC is to be deleted.

Reprocedure:

1. Create a single node cluster with [kind](https://github.com/kubernetes-sigs/kind).
2. Create a StatefulSet with the following manifests:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: test
  namespace: default
spec:
  clusterIP: None
  selector:
    app: test
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: test
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: test
  serviceName: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - command:
        - pause
        image: quay.io/cybozu/ubuntu:18.04
        name: ubuntu
        volumeMounts:
        - mountPath: /usr/share/nginx/html
          name: www
  volumeClaimTemplates:
  - metadata:
      name: www
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
      volumeMode: Filesystem
```

3. Once the pod gets running, edit PVC to add a finalizer `aaa.bbb/ccc`.
4. Run `kubectl delete --wait=false pvc/www-test-0`.
5. Run `kubectl delete --wait=false pods/test-0`.
6. See a new pod is created and become pending.
7. Edit the PVC to remove the `aaa.bbb/ccc` finalizer.
8. See the PVC is get deleted.
9. See the Pod remains pending forever.

**Does this PR introduce a user-facing change?**:
```release-note
Adding fix to statefulset controller to wait for pvc deletion before creating pods.
```

### CODE REVIEW

**Comment by ymmt2005** [comment]
@enisoc @kow3ns @tnozicka 
Hi, could you give an estimate when you can review this pull request?
Right now I want someone to give `/ok-to-test` to run tests.

**Comment by yuyulei** [comment]
https://github.com/kubernetes/kubernetes/pull/83566

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 1]
```diff
@@ -192,6 +192,8 @@ func (spc *realStatefulPodControl) createPersistentVolumeClaims(set *apps.Statef
 		case err != nil:
 			errs = append(errs, fmt.Errorf("failed to retrieve PVC %s: %s", claim.Name, err))
 			spc.recordClaimEvent("create", set, pod, &claim, err)
+		case !pvc.DeletionTimestamp.IsZero():
```
Comment (tnozicka) [inline pkg/controller/statefulset/stateful_pod_control.go]: DeletionTimestamp != nil

**Review by tnozicka — COMMENTED** [review]
/ok-to-test
(I can take closer look after PTO but at least I can start the tests here for you)

File: pkg/controller/statefulset/stateful_pod_control.go (hunk 1 above)
Comment (ymmt2005) [reply pkg/controller/statefulset/stateful_pod_control.go]: Fixed and rebased.  Thanks.

**Comment by ymmt2005** [comment]
Tests have passed.  PTAL @tnozicka

**Comment by cwdsuzhou** [comment]
Seems your PR is to solve the same issue with
https://github.com/kubernetes/kubernetes/pull/79164

**Comment by ymmt2005** [comment]
@cwdsuzhou 
Thanks.  Duplicate of #79164.  Hope it gets merged.

**Comment by ymmt2005** [comment]
@cwdsuzhou 
Hi.  Your PR #79164 was closed before got merged.
Do you know the current status of this bug?

**Comment by ymmt2005** [comment]
I've reopened this in #93457 because both #79164 and this were closed,
yet the problem still exists.

**Pull request closed without merging**